src/hf_tokenizer.cpp

```diff
@@ -249,7 +309,18 @@ Error HFTokenizer::_encode(
     const std::string& input,
     std::vector<uint64_t>& ret,
     uint64_t& last_piece_token_len) const {
-  for (const auto& piece : _pretokenizer->pre_tokenize(input)) {
+  // Apply normalization first if normalizer is available
+  std::string normalized_input = input;
+  if (_normalizer) {
+    normalized_input = _normalizer->normalize(input);
+    TK_LOG(
+        Info,
+        "normalized input: '%s' -> '%s'",
+        input.c_str(),
+        normalized_input.c_str());
+  }
+
+  for (const auto& piece : _pretokenizer->pre_tokenize(normalized_input)) {
     const auto result = token_map_->tryGetInteger(piece);
```

Comment:
Extra but can you help me add this

```suggestion
    # Check if the entire word is already a token to skip merging.
    const auto result = token_map_->tryGetInteger(piece);
```